src/drivers/src/USBDetection.cpp

```diff
@@ -0,0 +1,96 @@
+*/
+namespace USBDetection {
+
+    UsbDetector::UsbDetector() {
+        libusb_context *tmp;
+        libusb_init(&tmp);
+        ctx = ctx_t(tmp,libusb_exit);
+    }
+
+    std::set<usbdev_t> UsbDetector::find_arduinos() {
+        libusb_device **list;
+        cnt = libusb_get_device_list(ctx.get(), &list);
+
+        std::set<usbdev_t> arduinos;
+
+        int err = 0;
+        std::cout << cnt << std::endl;
+        if (cnt < 0) {
+            std::cerr << "error" << std::endl;
+        }
+        for (int i = 0; i < cnt; i++) {
+            struct libusb_device_descriptor info;
+            libusb_get_device_descriptor(list[i], &info);
+            if (info.idVendor == 0x2a03) { // 2a03 is arduino vendor id
+                libusb_ref_device(list[i]);
+                arduinos.insert(usbdev_t(list[i], libusb_unref_device));
+            }
+        }
+        libusb_free_device_list(list, 1);
+        return arduinos;
+    }
+
+    ArduinoConnection UsbDetector::open_arduino_with_prefix(std::string prefix) {
+        auto arduinos = find_arduinos();
+
+        size_t sz = arduinos.size();
+        std::vector<dhandle_t> handles(sz);
+        
+        int i = 0;
+        for (usbdev_t const &it : arduinos) {
+            libusb_device_handle *devhandle;
+            int err = libusb_open(it.get(), &devhandle);
+            if (err == 0) {
+                handles[i] = dhandle_t(devhandle, libusb_close);
+            } else {} //TODO
+            i++;
+        }
+
+        for (dhandle_t &it : handles) {
+            
+            libusb_device_handle *current_handle = it.get();
+
+            int numBytes;
+            unsigned char data[512];
+
+            int transfer = libusb_bulk_transfer(current_handle, LIBUSB_ENDPOINT_IN, data, sizeof(data), &numBytes, 0);
+
+            std::string dataStr((char *)data, 0, 10);
+            if (dataStr.compare(0, prefix.length(), prefix) == 0) {
+                return ArduinoConnection(std::move(it));
+            }
+        }
+        return ArduinoConnection();
+    }
+
+    ArduinoConnection UsbDetector::open_arduino_with_prefix(char *prefix) {
+        std::string prefixStr(prefix);
+        return open_arduino_with_prefix(prefixStr);
+    }
+
+
+    ArduinoConnection::ArduinoConnection(dhandle_t handle) {
+        arduino = std::move(handle); 
+    }
+
+    /*void ArduinoConnection::write(const std::string &msg) {
+        unsigned char *data = reinterpret_cast<unsigned char *>(const_cast<char *>((const_cast<std::string&>(msg)).c_str()));
+
+        ssize_t data_sz = msg.length();
+        int numBytes;
+        int transfer = libusb_bulk_transfer(arduino, LIBUSB_ENDPOINT_OUT, data, data_sz, &numBytes, 0);
+    }*/
+
+    std::string ArduinoConnection::read() {
+        unsigned char data[512];
+        int numBytes;
+        int transfer = libusb_bulk_transfer(arduino.get(), LIBUSB_ENDPOINT_IN, data, sizeof(data), &numBytes, 0);
+
+        return std::string(data, data+numBytes);
```

Comment:
Tricky note 1: this may not be the most useful return type. What if the user wants to read raw binary data that is not text? Unsigned-signed conversion will mess this up + the result will not be a meaningful string.
You could return a `std::vector<unsigned char>` to fix this issue.

Tricky note 2: the user can't ask for more or less than 512 bytes... this might work often, but in more complex cases will stop working correctly (you get get half a message or maybe no message at all, etc...).

This seems to be a complex issue, see http://www.beyondlogic.org/usbnutshell/usb4.shtml - do we really want to use bulk transfer here anyway? What kind of information do we expect to read? If you've already discussed this, then adding a comment may help future developers (and me). If not, this is the part where we think carefully about how we want to transfer data and what kind of data we want to transfer.